File: 13.罗马数字转整数.py
#
# @lc app=leetcode.cn id=13 lang=python
#
# [13] 罗马数字转整数
#

# @lc code=start
class Solution(object):
    def romanToInt(self, s):
        """
        :type s: str
        :rtype: int
        """
        dict_map = {"I":1, "V":5, "X":10, "L":50, "C":100, "D":500, "M":1000}
        sum = 0
        for i in range(len(s)):
            if i>0 and dict_map[s[i]]>dict_map[s[i-1]]:
                sum = sum + dict_map[s[i]] - 2 * dict_map[s[i-1]]
            else:
                sum = sum + dict_map[s[i]]
        return sum
# @lc code=end

File: 13.罗马数字转整数.java
import java.util.HashMap;

/*
 * @lc app=leetcode.cn id=13 lang=java
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
class Solution {
    public int romanToInt(String s) {
        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
        map.put('I', 1);
        map.put('V', 5);
        map.put('X', 10);
        map.put('L', 50);
        map.put('C', 100);
        map.put('D', 500);
        map.put('M', 1000);

        int[] nums = new int[s.length()];

        for (int i = 0; i < s.length(); i++) {
            nums[i] = map.get(s.charAt(i));
            if (i > 0 && nums[i] > nums[i - 1]) {
                nums[i - 1] = -nums[i - 1];
            }
        }
        int sum = 0;
        for (int i = 0; i < s.length(); i++) {
            sum = sum + nums[i];
        }
        return sum;
    }
}
// @lc code=end
